overlord/devicestate/handlers_install.go

```diff
@@ -920,16 +925,92 @@ func (m *DeviceManager) doFactoryResetRunSystem(t *state.Task, _ *tomb.Tomb) err
 		return fmt.Errorf("cannot use gadget: %v", err)
 	}
 
+	var trustedInstallObserver *boot.TrustedAssetsInstallObserver
+	// get a nice nil interface by default
+	var installObserver gadget.ContentObserver
+	trustedInstallObserver, err = boot.TrustedAssetsInstallObserverForModel(model, gadgetDir, useEncryption)
+	if err != nil && err != boot.ErrObserverNotApplicable {
+		return fmt.Errorf("cannot setup asset install observer: %v", err)
+	}
+	if err == nil {
+		installObserver = trustedInstallObserver
+		if !useEncryption {
+			// there will be no key sealing, so past the
+			// installation pass no other methods need to be called
+			trustedInstallObserver = nil
+		}
+	}
+
+	var installedSystem *install.InstalledSystemSideData
 	// run the create partition code
 	logger.Noticef("create and deploy partitions")
 	timings.Run(perfTimings, "factory-reset", "Factory reset", func(tm timings.Measurer) {
 		st.Unlock()
 		defer st.Lock()
-		_, err = installFactoryReset(model, gadgetDir, kernelDir, "", bopts, nil, tm)
+		installedSystem, err = installFactoryReset(model, gadgetDir, kernelDir, "", bopts, installObserver, tm)
 	})
 	if err != nil {
 		return fmt.Errorf("cannot perform factory reset: %v", err)
 	}
+	logger.Noticef("devs: %+v", installedSystem.DeviceForRole)
+
+	if trustedInstallObserver != nil {
+		// at this point we removed boot and data. sealed keys are becoming
+		// useless
+		err := os.Remove(filepath.Join(boot.InitramfsSeedEncryptionKeyDir, "ubuntu-data.recovery.sealed-key"))
+		if err != nil && !os.IsNotExist(err) {
+			return fmt.Errorf("cannot cleanup obsolete key file: %v", err)
+		}
+
+		// it is ok if the recovery key file on disk does not exist;
+		// ubuntu-save was opened during boot, so the removal operation
+		// can be authorized with a key from the keyring
+		err = secbootRemoveRecoveryKeys(map[secboot.RecoveryKeyDevice]string{
+			{Mountpoint: boot.InitramfsUbuntuSaveDir}: filepath.Join(boot.InstallHostFDEDataDir, "recovery.key"),
+		})
+		if err != nil {
+			return fmt.Errorf("cannot remove recovery key: %v", err)
+		}
+
+		// new encryption key for save
+		key, err := keys.NewEncryptionKey()
+		if err != nil {
+			return fmt.Errorf("cannot create encryption key: %v", err)
+		}
+
+		saveNode := installedSystem.DeviceForRole[gadget.SystemSave]
+		if saveNode == "" {
+			return fmt.Errorf("internal error: no system-save device")
+		}
+
+		if err := secbootStageEncryptionKeyChange(saveNode, key); err != nil {
+			return fmt.Errorf("cannot change encryption keys: %v", err)
+		}
+
+		installedSystem.KeyForRole[gadget.SystemSave] = key
+
+		// sanity check
+		if len(installedSystem.KeyForRole) == 0 || installedSystem.KeyForRole[gadget.SystemData] == nil || installedSystem.KeyForRole[gadget.SystemSave] == nil {
```

Comment:
we just set the entry for SystemSave a few lines above, shouldn't we check just for data at the beginning of thise large if stanza instead?